pkg/ruler/rule_concurrency.go

```diff
@@ -0,0 +1,213 @@
+func NewDynamicSemaphore(logger log.Logger, maxConcurrency func() int64) *DynamicSemaphore {
+	return &DynamicSemaphore{
+		logger:         logger,
+		maxConcurrency: maxConcurrency,
+		acquired:       0,
+	}
+}
+
+// TryAcquire tries to acquire a token from the semaphore.
+func (ds *DynamicSemaphore) TryAcquire() bool {
+	ds.mtx.Lock()
+	defer ds.mtx.Unlock()
+	maxTokens := ds.maxConcurrency()
+	if ds.acquired < maxTokens {
+		ds.acquired++
+		return true
+	}
+
+	return false
+}
+
+// Release releases a token back to the semaphore.
+func (ds *DynamicSemaphore) Release() {
+	ds.mtx.Lock()
+	defer ds.mtx.Unlock()
+	if ds.acquired > 0 {
+		ds.acquired--
+	} else {
+		panic("dynamic semaphore: release called without corresponding acquire")
+	}
+}
+
+type MultiTenantConcurrencyControllerMetrics struct {
+	CurrentConcurrency prometheus.Gauge
+	AcquireTotal       prometheus.Counter
+	AcquireFailedTotal prometheus.Counter
+	ReleaseTotal       prometheus.Counter
+}
+
+func newMultiTenantConcurrencyControllerMetrics(reg prometheus.Registerer) *MultiTenantConcurrencyControllerMetrics {
+	return &MultiTenantConcurrencyControllerMetrics{
+		CurrentConcurrency: promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+			Name: "cortex_ruler_global_rule_evaluation_concurrency_current",
+			Help: "Current number of active global concurrency slots",
+		}),
+		AcquireTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_ruler_global_rule_evaluation_concurrency_acquire_total",
+			Help: "Total number of acquired concurrency slots",
+		}),
+		AcquireFailedTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_ruler_global_rule_evaluation_concurrency_acquire_failed_total",
+			Help: "Total number of failed attempts to acquire concurrency slots",
+		}),
+		ReleaseTotal: promauto.With(reg).NewCounter(prometheus.CounterOpts{
+			Name: "cortex_ruler_global_rule_evaluation_concurrency_release_total",
+			Help: "Total number of released concurrency slots",
+		}),
+	}
+}
+
+// MultiTenantConcurrencyController is a concurrency controller that limits the number of concurrent rule evaluations both global and per tenant.
+type MultiTenantConcurrencyController struct {
+	logger               log.Logger
+	limits               RulesLimits
+	maxGlobalConcurrency int64
+	metrics              *MultiTenantConcurrencyControllerMetrics
+
+	globalConcurrency semaphore.Weighted
+	mtx               sync.Mutex
+	tenantConcurrency map[string]*DynamicSemaphore
+}
+
+// NewMultiTenantConcurrencyController creates a new MultiTenantConcurrencyController.
+func NewMultiTenantConcurrencyController(logger log.Logger, maxGlobalConcurrency int64, limits RulesLimits, reg prometheus.Registerer) *MultiTenantConcurrencyController {
+	return &MultiTenantConcurrencyController{
+		metrics:              newMultiTenantConcurrencyControllerMetrics(reg),
+		logger:               log.With(logger, "component", "ruler.MultiTenantConcurrencyController"),
```

Comment:
[nit] I would simplify it. Doesn't have to match the code struct name.

```suggestion
		logger:               log.With(logger, "component", "concurrency-controller"),
```

Reply:
Done in https://github.com/grafana/mimir/pull/8146/commits/004f3a11b2208bc5318a49460b1e10fc89d7e3bd (good to know, I followed this pattern because I saw others with this structure but good to know this is acceptable)